src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+// Manages ~/.fauna-shell
+
+import fs from "fs";
+import os from "os";
+import path from "path";
+const ini = require("ini");
+
+class InvalidConfigError extends Error {}
+
+// Wraps an `ini` file with helpers to get typed values out
+class Config {
+  keyName: string;
+  private config: { [key: string]: unknown };
+
+  static parse(keyName: string, data: string) {
+    return new Config(keyName, ini.parse(data));
+  }
+
+  constructor(keyName: string, config: { [key: string]: unknown }) {
+    this.keyName = keyName;
+    this.config = config;
+  }
+
+  strOpt(key: string): string | undefined {
+    const v = this.config[key];
+    if (v === undefined || typeof v === "string") {
+      return v;
+    } else {
+      throw new InvalidConfigError(
+        `Expected string for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  numberOpt(key: string): number | undefined {
+    const v = this.config[key];
+    if (v === undefined || typeof v === "number") {
+      return v;
+    } else if (typeof v === "string") {
+      try {
+        return parseInt(v);
+      } catch (_) {
+        throw new InvalidConfigError(
+          `Invalid number for ${this.keyName} ${key}`
+        );
+      }
+    } else {
+      throw new InvalidConfigError(
+        `Expected number for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  object(key: string): Config {
+    const v = this.config[key];
+    if (v !== null && typeof v === "object") {
+      return new Config(this.keyName, v as any);
+    } else {
+      throw new InvalidConfigError(
+        `Expected object for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  str(key: string): string {
+    return this.require(key, this.strOpt(key));
+  }
+
+  number(key: string): number {
+    return this.require(key, this.numberOpt(key));
+  }
+
+  require<T>(key: string, value: T | undefined): T {
+    if (value === undefined) {
+      throw new InvalidConfigError(
+        `Missing value for required ${this.keyName} ${key}`
+      );
+    } else {
+      return value;
+    }
+  }
+
+  entries(pred: (key: string) => boolean): [string, Config][] {
+    return Object.keys(this.config).flatMap((k) =>
+      pred(k) ? [[k, this.object(k)]] : []
+    );
+  }
+}
+
+// Represents `~/.fauna-shell`
+class RootConfig {
+  defaultEndpoint?: string;
+  endpoints: { [key: string]: Endpoint };
+
+  constructor(config: Config) {
+    this.defaultEndpoint = config.strOpt("default");
+    this.endpoints = Object.fromEntries(
+      config
+        .entries((k) => k !== "default")
+        .map(([k, v]) => [k, new Endpoint(v)])
+    );
+
+    if (this.defaultEndpoint === "default") {
```

Comment:
would this be like:

```
default = default

[default]
secret = blah

```
?

Would it fail to parse in that scenario?

Reply:
It would fail to parse. That file in JSON is like:
```json
{
  "default": "default",
  "default": {
    "secret": "blah",
  }
}
```